### PR TITLE
feat: favor ahash collections over std ones when applicable

### PR DESCRIPTION
**Pull request by hanabi1224**
## Summary of changes

Changes introduced in this pull request:
- favor ~~hashbrown~~ ahash collections over std ones when applicable
- as a side product, it caught the bug in https://github.com/ChainSafe/forest/pull/2453

## Reference issue to close (if applicable)

Closes https://github.com/ChainSafe/forest/issues/2441

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG](https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md) is up-to-date. All user-facing changes should be reflected in this document.

### CODE REVIEW

**Comment by lemmih** [comment]
Is this still relevant?

> Since Rust 1.36, this is now the HashMap implementation for the Rust standard library. However you may still want to use this crate instead since it works in environments without std, such as embedded systems and kernels.

**Comment by hanabi1224** [comment]
> How about we use ahash::HashMap?

@lemmih that also works except that the type name from ahash is slightly different, https://docs.rs/ahash/0.8.2/ahash/struct.AHashMap.html
would u prefer `use ahash::AHashMap as HashMap` or using `AHashMap` in place?

**Comment by lemmih** [comment]
Using `AHashMap` would also be fine by me.

```rust
use ahash::AHashMap;

let mut map: AHashMap<i32, i32> = AHashMap::new();
map.insert(12, 34);
```

**Comment by hanabi1224** [comment]
@lemmih Fixed. Using `use ahash::{HashMap, HashMapExt};` approach to reduce num of changed lines

File: blockchain/chain_sync/src/tipset_syncer.rs [hunk 1]
```diff
@@ -39,15 +22,31 @@ use forest_message::Message as MessageTrait;
 use forest_networks::Height;
 use forest_state_manager::Error as StateManagerError;
 use forest_state_manager::StateManager;
+use forest_utils::io::ProgressBar;
+use futures::stream::FuturesUnordered;
 use futures::Stream;
+use futures::StreamExt;
+use futures::TryFutureExt;
```
Comment (lemmih) [inline blockchain/chain_sync/src/tipset_syncer.rs:29]:
```suggestion
use futures::{Stream, StreamExt, TryFutureExt};
```

File: blockchain/message_pool/src/msgpool/msg_pool.rs [hunk 2]
```diff
@@ -16,6 +16,9 @@ use crate::msgpool::{republish_pending_messages, select_messages_for_block};
 use crate::msgpool::{RBF_DENOM, RBF_NUM};
 use crate::provider::Provider;
 use crate::utils::get_base_fee_lower_bound;
+use ahash::HashMapExt;
+use ahash::HashSetExt;
+use ahash::{HashMap, HashSet};
```
Comment (lemmih) [inline blockchain/message_pool/src/msgpool/msg_pool.rs:21]:
```suggestion
use ahash::{HashMap, HashMapExt, HashSet, HashSetExt};
```

**Review by lemmih — APPROVED** [review]
Looks good!

**Comment by LesnyRumcajs** [comment]
Do we see any measurable improvement in our case?

**Comment by hanabi1224** [comment]
> Do we see any measurable improvement in our case?

@LesnyRumcajs No and I doubt it will lead to any measurable improvement since the internal hashmap/hashset we use are quite small

**Comment by LesnyRumcajs** [comment]
> > Do we see any measurable improvement in our case?
> 
> @LesnyRumcajs No and I doubt it will lead to any measurable improvement since the internal hashmap/hashset we use are quite small

Is it worth the added dependency/mental overhead, then? I mean, I prefer to default to standard library collections unless there's a compelling reason to switch to something specialised.

**Comment by hanabi1224** [comment]
> unless there's a compelling reason to switch to something specialised.

@LesnyRumcajs the only motivation behind this is `AHash` being ~10x faster than `SipHash` as shown in the benchmark, if that's strong enough we could revisit this after getting performance measurements (but I doubt if we could see any non-trival gain, so it would be low priority)

**Review by LesnyRumcajs — APPROVED** [review]
Not a huge fan of this change, but I'm okay with it.

**Comment by hanabi1224** [comment]
@LesnyRumcajs I think your argument is reasonable, I'm kinda neutral about whether or not we should merge this.  @lemmih  what do you think?

**Comment by lemmih** [comment]
> @LesnyRumcajs I think your argument is reasonable, I'm kinda neutral about whether or not we should merge this. @lemmih what do you think?

Might as well merge it. PR is done and we were already using `AHashMap` in one place before. It might not matter for performance but I do like the consistency of always using either `ahash::HashMap` or the standard `HashMap`.

**Comment by LesnyRumcajs** [comment]
If we aim for not-using standard `HashMap` at all, should we create a `.clippy.toml` file with it disallowed like [here](https://github.com/amethyst/specs/blob/81073f340425803c99e3292fa14ceb00245bfea5/clippy.toml#L1)?

**Comment by lemmih** [comment]
> If we aim for not-using standard `HashMap` at all, should we create a `.clippy.toml` file with it disallowed like [here](https://github.com/amethyst/specs/blob/81073f340425803c99e3292fa14ceb00245bfea5/clippy.toml#L1)?

Will that work? We're still using `std::collections::HashMap`. We just instantiate it with a different hasher. Ie: `std::collections::HashMap<Key, Value, ahash::RandomState>`.

**Comment by hanabi1224** [comment]
I just tried `.clippy.toml`, it does not complain for `ahash::HashMap` but we do use `std::collections::HashMap` in some places where upstream crate API requires it

**Comment by LesnyRumcajs** [comment]
@hanabi1224 Thanks for checking it out. Let's just merge then.

**Comment by hanabi1224** [comment]
> @hanabi1224 Thanks for checking it out. Let's just merge then.

@LesnyRumcajs Actually I can explicitly allow those 2 instances and still add `.clippy.toml`, just pushed the change

**Pull request merged**